flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/PreciseTimestamp.java

```diff
@@ -0,0 +1,234 @@
+
+import org.apache.flink.table.types.logical.TimestampType;
+import org.apache.flink.util.Preconditions;
+
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.Arrays;
+
+import static org.apache.calcite.avatica.util.DateTimeUtils.timestampStringToUnixDate;
+import static org.apache.calcite.avatica.util.DateTimeUtils.unixTimestamp;
+import static org.apache.calcite.avatica.util.DateTimeUtils.unixTimestampToString;
+
+/**
+ * Immutable SQL TIMESTAMP type with nanosecond precision.
+ */
+public class PreciseTimestamp implements Comparable<PreciseTimestamp> {
+
+	public static final int MAX_PRECISION = TimestampType.MAX_PRECISION;
+
+	public static final int MIN_PRECISION = TimestampType.MIN_PRECISION;
+
+	private final long milliseconds;
+
+	private final int nanoseconds;
+
+	private final int precision;
+
+	private PreciseTimestamp(long milliseconds, int nanoseconds, int precision) {
+		Preconditions.checkArgument(precision >= MIN_PRECISION && precision <= MAX_PRECISION);
+		Preconditions.checkArgument(nanoseconds >= 0 && nanoseconds <= 999_999_999);
+		this.milliseconds = milliseconds;
+		this.nanoseconds = nanoseconds;
+		this.precision = precision;
+	}
+
+	@Override
+	public int compareTo(PreciseTimestamp that) {
+		return this.milliseconds == that.milliseconds ?
+			Integer.compare(this.nanoseconds, that.nanoseconds) :
+			Long.compare(this.milliseconds, that.milliseconds);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (!(obj instanceof PreciseTimestamp)) {
+			return false;
+		}
+		PreciseTimestamp that = (PreciseTimestamp) obj;
+		return this.compareTo(that) == 0;
+	}
+
+	@Override
+	public String toString() {
+		// yyyy-MM-dd HH:mm:ss
+		StringBuilder timestampString = new StringBuilder(unixTimestampToString(milliseconds));
+		if (precision > 0 && nanoseconds > 0) {
+			// append fraction part
+			timestampString.append('.').append(nonosSecondsToString(nanoseconds, precision));
+		}
+		return timestampString.toString();
+	}
+
+	public long toLong() {
+		return milliseconds;
+	}
+
+	public static PreciseTimestamp fromLong(long milliseconds, int precision) {
+		Timestamp t = new Timestamp(milliseconds);
+		long millis = isCompact(precision) ?
+			zeroLastNDigits(t.getTime(), 3 - precision) : t.getTime();
+		int nanos = (int) zeroLastNDigits((long) t.getNanos(), 9 - precision);
+		return new PreciseTimestamp(millis, nanos, precision);
+	}
+
+	public Timestamp toTimestamp() {
+		return Timestamp.valueOf(this.toString());
+	}
+
+	public static PreciseTimestamp fromTimestamp(Timestamp t, int precision) {
+		String timestampString = t.toString();
+		long millis = timestampStringToUnixDate(timestampString);
+		int nanos = timestampStringToNanoseconds(timestampString);
+		return new PreciseTimestamp(millis, nanos, precision);
+	}
+
+	public LocalDateTime toLocalDateTime() {
+		long epochSeconds = (milliseconds >= 0 || nanoseconds == 0) ?
```

Comment:
Why this check: `(milliseconds >= 0 || nanoseconds == 0)`?

Reply:
the nanoseconds should between 0-999_999_999.
so given a negative long (-123) as milliseconds, the epoch second would be -1 (not 0) and the nanoseconds would be (1000_000_000 - 123).

Reply:
Now, nanos has no mills now. It must be 0~999999, so it should never be negative.
We don't need to do everything same as `java.sql.Timestamp`, it has different usage scenarios and strange design.

Reply:
```
public LocalDateTime toLocalDateTime() {
		int date = (int) (milliseconds / MILLIS_PER_DAY);
		int time = (int) (milliseconds % MILLIS_PER_DAY);
		if (time < 0) {
			--date;
			time += MILLIS_PER_DAY;
		}
		int nanos = time * 1000_000 + nanosOfMills;
		LocalDate localDate = LocalDate.ofEpochDay(date);
		LocalTime localTime = LocalTime.ofNanoOfDay(nanos);
		return LocalDateTime.of(localDate, localTime);
	}
```
What do you think?